### DISCUSSION TITLE
Configuration validation fails in pipeline

### DISCUSSION THREAD

**Discussion by max06** [renovatebot/renovate#24551] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.97.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

We used a configuration file in a repository with the `json5`-suffix. Said file contains comments.

After renaming that file to `.json` (for proper syntax highlighting) the validation in a gitlab pipeline fails pointing to said comments.

```jsonc
WARN: Configuration Warning (repository=some-namespace/some-project)
       "configError": {
         "validationSource": "renovate.json",
         "validationError": "Invalid JSON (parsing failed)",
         "validationMessage": "Syntax error: expecting String near {\n  // Thi",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at checkForRepoConfigError (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/repository/init/merge.ts:205:17)\n    at mergeRenovateConfig (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/repository/init/merge.ts:229:3)\n    at getRepoConfig (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/repository/init/config.ts:12:12)\n    at initRepo (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/repository/init/index.ts:44:12)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/repository/index.ts:55:14)\n    at attributes.repository (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/global/index.ts:1[84](https://somegitlabserver/another-orga/andnested/project/-/jobs/3910709#L84):11)\n    at start (/opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/36.97.1/node_modules/renovate/lib/renovate.ts:18:22"
       },
       "res": "created"
```

Following partial configuration file is used: 

```jsonc
{
  // This file is in JSON5 format: https://json5.org
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "ignorePaths": [
    "**/rendered/**"
  ],
  "regexManagers": [
    {
...
```

The validator on my local machine (`36.97.3` via npm) is happy about the config, the pipeline isn't.

Is jsonc not supported? Is the different result of the validation a bug?

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
You should first try to understand why you get different results locally vs in the pipeline

**Reply by max06** [reply]
Uhm... same configuration file, different results? Sounds like a bug to me...

**Reply by viceice** [reply]
probably different renovate versions?

**Reply by max06** [reply]
Minor changes in between. The pipeline version obviously has issues with the comments in the .json-file, while comments in the .json5-version (same content!) are perfectly fine. The cli doesn't care about comments.

**Reply by max06** [reply]
Just in case my report is unclear: The difference is the configuration file suffix. Renovate is happy *with* comments in the `renovate.json5`-file, but complaints if I rename the file to `renovate.json`. The validator doesn't care about the suffix at all.

**Reply by rarkins** [reply]
When you said "same configuration file" you meant same contents, and ignore that the filename extension is different?

**Reply by max06** [reply]
>We used a configuration file in a repository with the json5-suffix. Said file contains comments.
>
>After renaming that file to .json ...

No idea how I could make it more clear...

**Reply by rarkins** [reply]
Wait. It also has some minor contents changes too? Are you sure that you meant to use the word "same" if both the filename/extension and contents were different?

**Reply by max06** [reply]
There are no content changes. Only the file name suffix.

**Reply by max06** [reply]
Alright... one step back.

Renovate in pipeline: `36.97.1`
Validator from npm: `36.97.3` (that's the minor difference)

Renovate in pipeline takes:
- json with comments when suffix is `json5`
- json without comments when suffix is `json`

Renovate in pipeline does not take 
- json with comments when suffix is `json`

The contents described as "json with comments" are identical in both files. The file was simply renamed. 


Renovate config validation on localhost takes:
- json with comments when suffix is `json5`
- json with comments when suffix is `json`

The file is the same as used in the pipeline.

**Comment by rarkins** [comment]
There appears to be some mixing of json/json5 logic and potentially inconsistency of runtime validation versus the CLI validation. I have create this issue to address it: #24554

What this means in your case is that we won't do any change now to loosen the parsing of `.json` files in all circumstances, and you need to stick with `.json5` config files if you want to use JSON5 or JSONC syntax.

I think as this discussion illustrates, there's too much potential for confusion when `.json` files are allowed to violate JSON syntax.

**Reply by max06** [reply]
Thank you.

As additional information: It's not uncommon for tools to allow jsonc-syntax in regular json files - take vscode as prominent example. There is no actual violation of the json syntax, it's just relaxing the parser to ignore lines starting with `//`. Everything else defining json is still required as usual. I'd like to see support for jsonc in renovate, but I can understand your position here.

**Reply by rarkins** [reply]
Thanks for the additional context. Which tools don't properly syntax highlight .json5 then?

**Reply by max06** [reply]
Well... vscode would be the best example here 😂 

(I have never even heard of json5 before touching renovate, in my opinion it shouldn't exist at all.)

**Reply by rarkins** [reply]
So you can put a JSON5/JSONC comment into a `.json` file and VSCode will syntax highlight it correctly, but if you open a `.json5` file then it won't? Strange design choice if so!

When you say it shouldn't exist, do you mean:
1. JSON is perfect, nobody needs comments or trailing commas, or
2. JSONC is enough (comments?)
3. JSON5 shouldn't _need_ to exist - JSON should have supported comments/commas/etc?

**Reply by rarkins** [reply]
This discussion is very old, I wonder if it's still the latest status: https://github.com/Microsoft/vscode/issues/9464

i.e. can be solved with config, or a 7 year old extension

**Reply by max06** [reply]
No, actually not.

By default, vscode opens .json-files as standard json. As a user you have the option to change the selected language to jsonc without any extensions. Vscode doesn't want to lead users to creating jsonc files when the tools they're written for can't read them.

Vscodes own configuration files (all json) are treated as jsonc-files. You can use comments in them and vscode will not complain and render them properly. And as you said, there is no integrated support for json5 at all, without extensions.

My suggestion is: accept as much as possible, render only what's needed. If we're talking about configuration files, treat json files as jsonc. If a user doesn't use comments, you're still good. If we're talking about communication between tools, use strict json. 

By shouldn't exist I mean json5 shouldn't have been created in the first place. There are json and yaml as serializable and interchangeable formats. Json is great for exchanging data with everything, yaml is great for readability while still being machine readable. Yaml allows comments, jsonc got added for that reason. Adding another json variant with focus on readability by sacrificing interchangeability is the wrong way to go, in my personal opinion. 

As json5.org states:
>  It is not intended to be used for machine-to-machine communication. (Keep using JSON or other file formats for that. 🙂)

As said, personal opinion. I don't see any use in it.
